File: 278_First Bad Version.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        int version = 0;
        while (left < right){
            version = left+(right - left)/2;
            if (isBadVersion(version)){
                right = version; 
            }else{
                left = version + 1;
            }
        }
        return left;
        
    }
}

File: 278_First Bad Version.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int l = 1;
        int r = n;
        while(l < r){
            int mid = l + (r - l) / 2;
            if(!isBadVersion(mid) && isBadVersion(mid + 1))
                return mid + 1;
            if(!isBadVersion(mid))
                l = mid;
            else
                r = mid;
        }
        return 1;
    }
};
